### ISSUE TITLE
icingacli equivalent of "Business Impact" in the BP web module

### ISSUE THREAD

**Issue by bw00sy** [Icinga/icingaweb2-module-businessprocess#238]
I've posted something like this on the community but feel it's maybe a feature that could be added to the icingacli command for the business process as an enhancement. Please forgive me for posting there and here but I've started working a script to analyse the BP conf file logic but feel I'm reinventing the wheel. 

My request is ..would it be possible to have the icingacli command for this module to return the business impact of a host/service? Ideally a command like "icingacli businessprocess process check --host SERVER_1 --service CPU" returning the text equivalent of the Business Impact on the web actions. https://127.0.0.1/icingaweb2/businessprocess/node/impact?name=SERVER_1;CPU

For example...
SQL[Critical]:Server[Critical]:SERVER_1[Critical]:CPU[Critical]
Applications[Critical]:Sales[Critical]:SERVER_1[Critical]:CPU[Critical]

The reason for this request is the shortcomings of the icingacli-businessprocess check command in processing alerts and creating notifications. It cannot be relied upon to create a specific alert if another host or service fails in and already failed Business Process. Also there appear to be inconsistencies when acknowledging alerts and the process state changing for new failures. 

Basically I feel the logic on alerting would be better served by not sending a notification for the BP Check command directly but rather determining which Business Process, if any, is affected by a host or service state change. Additionally the full service output is available via the host/service alert yet there is nothing via the state change in a BP check command.

The code to write a notification script would be a lot simpler if I could determine which BP's an alert is affecting, or not.

**Comment by nilmerg** [comment]
Hi,

I don't quite get behind the idea yet.

> The reason for this request is the shortcomings of the icingacli-businessprocess check command in processing alerts and creating notifications. It cannot be relied upon to create a specific alert if another host or service fails in and already failed Business Process. 

Because the check only is interested in the process' state. If a node's state change doesn't affect the process' state the check reports nothing. Though, you should have a check for each separate node as well either way. That can be relied upon very well.

> Also there appear to be inconsistencies when acknowledging alerts and the process state changing for new failures.

Please be more elaborate here. We have another customer who made a request in this regard. They want the check to consider downtimes/acknowledgements of involved nodes of a process. Maybe that's the same for you and making the check aware of this also solves your problem?

> Basically I feel the logic on alerting would be better served by not sending a notification for the BP Check command directly but rather determining which Business Process, if any, is affected by a host or service state change.

In which way differs this? The result is still an alert that a business process is problematic. Also, if you do this, you have to re-invent another wheel: The consideration of the many operators a business process may use. (That's not just `AND` :wink: ) Honestly, the check command already does it's job pretty well.

> Additionally the full service output is available via the host/service alert yet there is nothing via the state change in a BP check command.

Because it's not its task. The BP check command only tells you that your entire shop infrastructure is broken. That's of interest to your sales manager (or whoever) to allow him to know that you've messed up. A separate check for each service should exist to tell **you** that there's something messed up. (Possibly a bit earlier so that the sale's manager doesn't even notice anything :wink: )

**Comment by bw00sy** [comment]
Hi,

Thanks for the reply. 

The request I have from clients is to limit the number of mails sent and **only** to use BP logic in alerting. It's becoming an increasingly important feature of monitoring NOT to have the old "HP Openview" logic in monitoring. I'd rather drill down through a failed BP to find the problem than look at a little red dot on a map and then wonder what it's impact is. This is why I feel the Business Process Module is the future of monitoring. Ultimately sending a notification of a failed BP status with host/service in the subject, and the details in the body of the message.

I'll try explain the logic for requesting the original enhancement. 

Let's call our BP "Shop". 

Server1 in Shop has CPU at 100% and this causes Shop to go Critical creating a Notification that Shop is now Critical. The sales manager/support guy sees this via my notification script which uses icingacli to show the BP tree with Server1 CPU critical. To prevent 2 alerts per failure, the individual service notification is **skipped** because it's already in a BP (Shop). This is what I do currently.

The current problem is......before the support guy can do anything about Server1's CPU, Server2 in Shop runs out of disk space. No BP notification is sent because it's state has not **changed**. Even acknowledging the original Server1 CPU status when it happened  **still** will not create a new BP notification for Shop as it's status is already critical. Setting it's "check execution" to volatile doesn't fix this obviously.

If I have the individual services send their notifications instead of the way I'm trying to do it, the sales managers/support guys have no idea of the business impact of Server2's disk space failure. Let's say they are nowhere near an icingaweb interface to check. The logic in my original request will solve this problem altogether. I would use the host/service notification, which could constantly change state, and send the notification show the affected BP.

I'm open to any suggestions to change my way of thinking :-). I'm merely trying to get people to pay attention to the alerts generated. Creating an alert that the "Coffee Processing" business process is not working and this is the reason why, holds more weight than saying Server_A is broken and if you're not at a workstation you'll have no idea of that impact. Till you go try make a cup of coffee ;-)

**Comment by nilmerg** [comment]
> Ultimately sending a notification of a failed BP status with host/service in the subject, and the details in the body of the message.

Got it.

> The current problem is......before the support guy can do anything about Server1's CPU, Server2 in Shop runs out of disk space. No BP notification is sent because it's state has not changed.

Though, that's not a problem of the business process check. It's one of your business process. It should have not been critical up until Server 2 ran out of disk space. At least if we're speaking here about a fail-over scenario. (`OR` condition)

> The logic in my original request will solve this problem altogether. I would use the host/service notification, which could constantly change state, and send the notification show the affected BP.

Still, doing so would require you to not only duplicate (at the moment) the logic to detect the impact but also the correct interpretation of the node's role/priority in each process. Not worth the effort, I'd rather extend the module so that the check already provides everything required.

--

I completely understand your reasoning to not want look at individual services first but at the their business process. But rest assured, your processes also need to be properly setup for this to work. So by enhancing the check command to focus more on the node that's causing the incident and showing a more complete picture of it's impact  should suffice to fulfill your client's request. (Though, still only for the respective process, other processes should have their own check/alert)

**Comment by bw00sy** [comment]
Ok, so my real world example:-

I monitor host resources (CPU, Memory, Disk); file queues; SQL queues, interface statistics; transactions and other custom services under a business process. **All** of them are sub nodes under one Business Process or another. Every monitored HOST/SERVICE is in some Business Process, be it the company chat server or something more critical.

Once **one** of those services fail, I will get a notification via the bp check command. None of the sub nodes are **|OR|** conditions as they **all** affect the running of the real world business process. There is no fail-over, it's pre-emptive. That is the way we designed the processes. For example: Interface errors **|AND|** CPU could both affect the time it takes to run a transaction at the point of sale interface.

With the current business process module, no **new** notification will be sent if an **ADDITIONAL** service fails for the same business process, unless the process changes state in the interim. Acknowledging existing alerts will not change this situation either. Ultimately, if it did send a new notification for the new failed check, I would use the bp check command for all notifications, and the original request would not exist.

> Still, doing so would require you to not only duplicate (at the moment) the logic to detect the impact but also the correct interpretation of the node's role/priority in each process

It's already being done in the "Business Impact" web portion of this business process module. All I was asking was for was to put the same logic in the cli.

**Comment by nilmerg** [comment]
As I said, I understand your reasoning. Though, the business process module is not the correct place to look at to solve this. Also, we won't add functionality that's only of interest to third-party products, especially if it's only a single (very specific) use-case.

> Ultimately, if it did send a new notification for the new failed check, I would use the bp check command for all notifications, and the original request would not exist.

This may change. As I already said as well, we have a customer who requested exactly this. The check command will consider handled states. So if there's an acknowledged problem and a new node has a problem a state change occurs and a new notification is sent.

So, while we won't add the functionality you want to the CLI, there's a chance it is not even necessary anymore.
